File: DynamicProgramming/CoinChange2/coinChange2.py
class CoinChange2(object):

    def __init__(self):
        self.track = []
        self.path = set()
        self.res = 0

    def backtrack(self, coins, rest):
        if rest == 0:
            self.track.sort()
            if tuple(self.track) not in self.path:
                self.res += 1
                self.path.add(tuple(self.track))
            return

        for i in range(len(coins)):
            if rest - coins[i] >= 0:
                rest = rest - coins[i]
                self.track.append(coins[i])
                self.backtrack(coins, rest)
                rest = rest + coins[i]
                self.track.pop(-1)
            else:
                continue

    def CoinChange2(self, coins, target):
        dp = [[0]*(target + 1) for i in range(len(coins) + 1)]
        for i in range(target + 1):
            dp[0][i] = 0
        for i in range(len(coins) + 1):
            dp[i][0] = 1

        for i in range(1, len(coins) + 1):
            for j in range(1, target + 1):
                if j - coins[i - 1] >= 0:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]]
                else:
                    dp[i][j] = dp[i - 1][j]
        
        return dp[len(coins)][target]

cc2 = CoinChange2()
coins = [1, 2, 5]
target = 5
print(cc2.CoinChange2(coins, target))
cc2.backtrack(coins, target)
print(cc2.res)

File: DynamicProgramming/CoinChange2/coinChange2.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.HashSet;

class CoinChange2 {
  public int res = 0;
  public List<Integer> track = new ArrayList<>();
  public HashSet<List<Integer>> path = new HashSet<>();

  public void backtrack(int[] coins, int reset) {
    if (reset == 0) {
      Collections.sort(track);
      if (!path.contains(track)) {
        path.add(track);
        System.out.println(track);
        this.res += 1;
      }
      return;
    }

    for (int i = 0; i < coins.length; i++) {
      if (reset - coins[i] >= 0) {
        track.add(coins[i]);
        reset = reset - coins[i];
        this.backtrack(coins, reset);
        track.remove(track.size() - 1);
        reset = reset + coins[i];
      } 
      else {
        continue;
      }
    }
  }

  public int coinChange2(int[] coins, int amount) {
    int[][] dp = new int[coins.length + 1][amount + 1];
    for (int i = 0; i < amount + 1; i++) {
      dp[0][i] = 0;
    }

    for (int i = 0; i < coins.length + 1; i++) {
      dp[i][0] = 1;
    }

    for (int i = 1; i < coins.length + 1; i++) {
      for (int j = 1; j < amount + 1; j++) {
        if (j - coins[i - 1] >= 0) {
          dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]];
        } else {
          dp[i][j] = dp[i - 1][j];
        }
      }
    }

    return dp[coins.length][amount];
  }

  public static void main(String[] args) {
    CoinChange2 cc2 = new CoinChange2();
    int[] coins = {1, 2, 5};
    int amount = 5;
    int res = cc2.coinChange2(coins, amount);
    System.out.println(res);
    cc2.backtrack(coins, amount);
    System.out.println(cc2.res);
  }
}